database/sqlserver/sqlserver.go

```diff
@@ -116,16 +119,43 @@ func WithInstance(instance *sql.DB, config *Config) (database.Driver, error) {
 	return ss, nil
 }
 
-// Open a connection to the database
+// Open a connection to the database.
 func (ss *SQLServer) Open(url string) (database.Driver, error) {
 	purl, err := nurl.Parse(url)
 	if err != nil {
 		return nil, err
 	}
 
-	db, err := sql.Open("sqlserver", migrate.FilterCustomQuery(purl).String())
-	if err != nil {
-		return nil, err
+	useMsiParam := purl.Query().Get("useMsi")
+	useMsi := false
+	if len(useMsiParam) > 0 {
+		useMsi, err = strconv.ParseBool(useMsiParam)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	var db *sql.DB
+	if useMsi {
+		tokenProvider, err := getMSITokenProvider(fmt.Sprintf("%s%s", "https://", strings.Join(strings.Split(purl.Hostname(), ".")[1:], ".")))
+		if err != nil {
+			return nil, err
+		}
+
+		connector, err := mssql.NewAccessTokenConnector(
+			migrate.FilterCustomQuery(purl).String(), tokenProvider)
+
+		if err != nil {
+			return nil, err
+		}
+
+		db = sql.OpenDB(connector)
```

Comment:
What's the difference between this `sql.OpenDB()` call and the below `sql.Open()` call? What's the reason for using two different methods to open the database?

Reply:
sql.Open will actually end up calling sql.OpenDB behind the scenes, but it builds the connector first based on what's passed in/the db context. sql.OpenDB just expects a connector, which is this case we're building ourselves, so we can call sql.OpenDB directly in that case. This follows the mssql guidance for using msi connections: [mssql msi](https://github.com/denisenkom/go-mssqldb#azure-active-directory-authentication---preview)

source code which has sql.Open and sql.OpenDB (sorry, can't permalink to sql.Open): https://golang.org/src/database/sql/sql.go

Reply:
Awesome, thanks for the explanation!